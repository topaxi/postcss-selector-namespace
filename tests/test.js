import path from 'path'
import { transformSnapshot, compareFixture, expectUnchanged } from './helpers'

describe('Basic functionality', () => {
  it('should work', () => {
    compareFixture('basic.css', {
      selfSelector: /:--component/,
      namespace: '.namespaced',
    })
  })

  it('has a default namespace selector of :--namespace', () => {
    transformSnapshot(':--namespace {}', { namespace: '.my-component' })
  })

  it('has a default namespace .self', () => {
    transformSnapshot(':--namespace {}')
  })

  it('works with a regexp which matches multiple selectors', () => {
    compareFixture('multiself.css', {
      selfSelector: /&|:--component/,
      namespace: '.my-component',
    })
  })

  it('works with :--namespace not being the first selector', () => {
    transformSnapshot('.foo :--namespace {}', {
      namespace: '.my-component',
    })
  })

  it('can accept a function for the namespace option', () => {
    transformSnapshot(
      '.foo {}',
      { namespace: file => `.${path.basename(file, '.css')}` },
      { from: 'bar.css' },
    )
  })

  it('does not apply if computed namespace is falsy', () => {
    transformSnapshot('.foo {}', { namespace: file => !!file })
  })

  it('works with multiple namespaces', () => {
    transformSnapshot('.selector {}', { namespace: '.foo, .bar' })
  })
})

describe(':root', () => {
  it('is configurable', () => {
    transformSnapshot(':root .foo {}:global .foo {}', {
      namespace: '.my-component',
      rootSelector: ':global',
    })
  })

  it('does not namespace :root selectors', () => {
    compareFixture('root.css', {
      selfSelector: /:--component/,
      namespace: '.my-component',
    })
  })

  it('does namespace :root selectors if ignoreRoot is false', () => {
    transformSnapshot(':root .foo {}', {
      namespace: '.my-component',
      ignoreRoot: false,
    })
  })

  it('does drop :root if dropRoot is true', () => {
    transformSnapshot(':root .foo {}', {
      namespace: '.my-component',
      dropRoot: false,
    })
  })

  it("does not drop :root if it's the only selector", () => {
    transformSnapshot(':root {}', {
      namespace: '.my-component',
      dropRoot: true,
    })
  })
})

describe('@keyframes', () => {
  it('does not transform keyframe definitions', () => {
    expectUnchanged(
      '@keyframes fadeout { from { opacity: 1 } to { opacity: 0 }}',
    )
  })

  it('does not transform vendor prefixed keyframe definitions', () => {
    expectUnchanged(
      '@-moz-keyframes fadeout { from { opacity: 1 } to { opacity: 0 }}',
    )
  })
})

describe('@supports', () => {
  it('does namespace in supports atrule', () => {
    transformSnapshot(
      '@supports (display: flex) { .bar { display: flex; } }',
      { namespace: '.my-component' },
    )
  })
})

describe('SCSS', function() {
  const syntax = require('postcss-scss')

  it('does transform basic nesting', () => {
    transformSnapshot(
      `
        & { .bar { color: red; } }
        .foo { color: blue }
      `,
      { selfSelector: '&', namespace: '.my-component' },
    )
  })

  it('does work with single line comments', () => {
    transformSnapshot(
      `
        & { .bar { color: red; } }
        //.foo { color: blue }
      `,
      { selfSelector: '&', namespace: '.my-component' },
      { syntax },
    )
  })

  it('does work with rules nested in nested media queries', () => {
    compareFixture(
      'nested-media-queries.scss',
      { namespace: '.my-component' },
      { syntax },
    )
  })

  it('works with :root selector', () => {
    transformSnapshot(':root { &.bar { color: red; } }', {
      selfSelector: '&',
      namespace: '.my-component',
      dropRoot: true,
    })
  })

  describe('@media nesting', () => {
    it('media with nested around properties', () => {
      transformSnapshot('& { .bar { @media (screen) { color: red; } } }', {
        selfSelector: '&',
        namespace: '.my-component',
      })
    })

    it('media with nested around selector', () => {
      transformSnapshot('& { @media (screen) { .bar { color: red; } } }', {
        selfSelector: '&',
        namespace: '.my-component',
      })
    })

    it('media with nested around namespaced selector', () => {
      transformSnapshot('@media (screen) { & { .bar { color: red; } } }', {
        selfSelector: '&',
        namespace: '.my-component',
      })
    })
  })

  describe('@include mixins', () => {
    it('works with pseudo elements', () => {
      transformSnapshot(
        `
          @mixin do-a-thing-with-pseudo-elements() {
            position: relative;
            &::before {
              content: '';
            }
          }
          .some-class {
            @include do-a-thing-with-pseudo-elements();
          }
        `,
        { selfSelector: '&', namespace: '.my-component' },
      )
    })
  })

  describe('@for', () => {
    it('does work with @for and nested @for rules', () => {
      compareFixture(
        'for.scss',
        { namespace: '.my-component', selfSelector: '&' },
        { syntax },
      )
    })
  })
})
