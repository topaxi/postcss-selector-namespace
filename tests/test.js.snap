// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`:root does drop :root if dropRoot is true 1`] = `":root .foo {}"`;

exports[`:root does namespace :root selectors if ignoreRoot is false 1`] = `".my-component :root .foo {}"`;

exports[`:root does not drop :root if it's the only selector 1`] = `":root {}"`;

exports[`:root does not namespace :root selectors 1`] = `
":root .my-component {
  color: red;
}

.foo {
  color: blue;
}
"
`;

exports[`:root is configurable 1`] = `".my-component :root .foo {}.foo {}"`;

exports[`@supports does namespace in supports atrule 1`] = `"@supports (display: flex) { .my-component .bar { display: flex; } }"`;

exports[`Basic functionality can accept a function for the namespace option 1`] = `".bar .foo {}"`;

exports[`Basic functionality does not apply if computed namespace is falsy 1`] = `".foo {}"`;

exports[`Basic functionality has a default namespace .self 1`] = `".self {}"`;

exports[`Basic functionality has a default namespace selector of :--namespace 1`] = `".my-component {}"`;

exports[`Basic functionality should work 1`] = `
"/* Toplevel pseudo */
.namespaced :hover {
  color: red;
}

/* Custom selector */
.namespaced {
  color: blue;
}

/* toplevel class */
.namespaced .title {
  font-weight: bold;
}

/* Custom */
.namespaced .title {
  color: black;
}

/* Toplevel element */
.namespaced div {
  display: block;
}

/* Multiple selectors */
.namespaced .foo, .namespaced .bar {
  color: yellow;
}

/* NOT */
.namespaced :not(.foo, .bar) {
  color: purple;
}
"
`;

exports[`Basic functionality works with :--namespace not being the first selector 1`] = `".foo .my-component {}"`;

exports[`Basic functionality works with a regexp which matches multiple selectors 1`] = `
".my-component {
  color: red;
}

.my-component {
  color: blue;
}
"
`;

exports[`Basic functionality works with multiple namespaces 1`] = `".foo .selector, .bar .selector {}"`;

exports[`SCSS @for does work with @for and nested @for rules 1`] = `
".my-component .pony {
  @for $i from 1 through 3 {
    &.test-#{$i} {
      color: pink;
      background-color: white;
    }
  }
}

@for $i from 1 through 3 {
  .my-component.column-#{$i} {
    display: flex;
  }
}
"
`;

exports[`SCSS @include mixins works with pseudo elements 1`] = `
"
          @mixin do-a-thing-with-pseudo-elements() {
            position: relative;
            &::before {
              content: '';
            }
          }
          .my-component .some-class {
            @include do-a-thing-with-pseudo-elements();
          }
        "
`;

exports[`SCSS @media nesting media with nested around namespaced selector 1`] = `"@media (screen) { .my-component { .bar { color: red; } } }"`;

exports[`SCSS @media nesting media with nested around properties 1`] = `".my-component { .bar { @media (screen) { color: red; } } }"`;

exports[`SCSS @media nesting media with nested around selector 1`] = `".my-component { @media (screen) { .bar { color: red; } } }"`;

exports[`SCSS does transform basic nesting 1`] = `
"
        .my-component { .bar { color: red; } }
        .my-component .foo { color: blue }
      "
`;

exports[`SCSS does work with rules nested in nested media queries 1`] = `
"$break-small: 320px;
$break-large: 1200px;

.my-component {
  float: left;
  width: 250px;
  h1 {
    font-weight: bold;
  }
  @media screen and (max-width: $break-small) {
    width: 100px;
    float: none;
    h1 {
      font-size: 32px;
    }
  }
  @media screen and (min-width: $break-large) {
    float: right;
    h1 {
      font-size: 48px;
    }
  }
}
"
`;

exports[`SCSS does work with single line comments 1`] = `
"
        .my-component { .bar { color: red; } }
        //.foo { color: blue }
      "
`;

exports[`SCSS works with :root selector 1`] = `":root { &.bar { color: red; } }"`;
